iTwin/iTwinUI
# `LabeledInput` API and styling improvements

Since this PR changes `LabeledInput`'s internal working, this is a pre-requisite for the `LabeledInput` related PR #1777.

## Motivation



This PR handles some `LabeledInput` API and styling changes based on the suggestions in https://github.com/iTwin/iTwinUI-migration-tool/pull/96#discussion_r1436581272:

1. When `StatusMessage`'s `svgIcon=null`, no icon is displayed along with the message (https://github.com/iTwin/iTwinUI/pull/1753/commits/fd663c0797512efe66f3069c76ca4802600c5628)
> it appears like this, but i'm thinking we should change the source code to auto-apply the fill too.
![](https://github.com/iTwin/iTwinUI-migration-tool/assets/9084735/7594975f-7ea5-4639-ba71-dae4c5b832b2)
2. Apply status fill to `LabeledInput`'s custom icon also (https://github.com/iTwin/iTwinUI/pull/1753/commits/6c387d5055186a182dbf017d0a5b62de30800f71)
> since we only support basic svg icons to be passed into the `svgIcon` prop, we should probably _not_ use `InputWithDecorations` inside `LabeledInput`/`LabeledTextarea` in v3.
`InputWithDecorations` should be reserved for advanced cases (like passing buttons).
3. Replaced `InputWithDecorations` with `InputFlexContainer` in `LabeledInput` and `LabeledTextarea` (https://github.com/iTwin/iTwinUI/pull/1753/commits/b2fe1397528082d0adbd8ece8121e0a4b11eacff)

## Changes

1. Addressed the points in the motivation section.
2. `.iui-input-with-icon` now does not add `padding-inline-end` when no end-icon is present.
3. Removed unnecessary `status` passed to `Select` and `Input` in `LabeledSelect` & `LabeledInput` respectively.
4. Removed accidental styling support for the invalid case of `.iui-input` and `.iui-input-with-icon > .iui-select-button` with `[data-iui-status='informational']`. This is invalid because an `informational` status on our input or select adds little to no meaning.
5. Deprecated `svgIcon` in `LabeledSelect` in favor of passing `<StatusMessage startIcon={svgIcon} />` to the `message` prop.
6. Updated `input.html` and `select.html` to visual test cases for adding `data-iui-status` on `iui-input-grid`, `iui-input`, and `iui-select-button`,
7. Add missing `@include iui-text-selection($status);` for the `.iui-input-flex-container[data-iui-status={…}] .iui-input` case.
8. Removed duplicate status styling applied on for the case where `data-iui-status` is applied on `.iui-input-grid` and the case where it is applied on `.iui-input, .iui-input-with-icon > .iui-select-button`.
9. Added missing `data-iui-status` for all uses of `<InputGrid>` in itwinui-react, since this data attribute exists in itwinui-css.

## Testing



* Motivation point 1: Added unit test in `StatusMessage.test.tsx` to confirm that no icon is in the DOM if `svgIcon=null`
* Motivation point 2: Added new "With Svg Icon" stories to `LabeledInput` and `LabeledTextarea` to test that the status fill is applied to `svgIcon`. Open to suggestions for better ways to test this.
* Motivation point 3: Edited failing unit tests to account for this change.
* Change point 2: The input with type number in the Table → Filter [story](https://github.com/iTwin/iTwinUI/pull/1753/files#diff-368e5b17d4aba72dcf8f6f5c2bbbd2bab587ac49076302931a5ca931167dbd38) tests this.
* Change point 3: No test images failed and the focus outlines appeared correctly with manual testing
* Change point 4: Mayank and I manually verified that adding an informational status on an input or select conveyed little to no extra meaning than without the status.
* Change point 5: Updated `LabeledSelect` story to discontinue using the newly deprecated `svgIcon` prop
* Added react-workshop tests on existing stories to test that the status focus outline is applied correctly.

## Docs



Added changesets

## TODOs after PR:

* [ ] Look into having a consistent padding among `input-flex-container`, `input-with-icon`, `search-box` (https://github.com/iTwin/iTwinUI/pull/1753#discussion_r1449405834)
  * Maybe create a new issue (or add a bullet point to an existing) issue
* [ ] Fix `InputGroup` bug: passing `<StatusMessage startIcon={…} />` to the message of `<InputGroup status="…" />` shows duplicate status icons. (https://github.com/iTwin/iTwinUI/pull/1753#discussion_r1459238732)
